src/server/metadata/cmds/cmds.go

```diff
@@ -37,3 +46,82 @@ func Cmds(mainCtx context.Context, pachctlCfg *pachctl.Config) []*cobra.Command
 
 	return commands
 }
+
+func parseKV(data string) (k, v string) {
+	// TODO(jrockway): Allow keys and values with = in them.
+	parts := strings.SplitN(data, "=", 2)
+	return parts[0], parts[1]
+}
+
+func parseData(data string) (result map[string]string, _ error) {
+	if err := json.Unmarshal([]byte(data), &result); err != nil {
+		return nil, errors.Wrap(err, "parse json data")
+	}
+	return
+}
+
+func parseEditMetadataCmdline(args []string) (*metadata.EditMetadataRequest, error) {
+	if len(args)%4 != 0 {
+		return nil, errors.New("must supply arguments in multiples of 4")
+	}
+	var errs error
+	result := &metadata.EditMetadataRequest{}
+	for i := 0; len(args) > 0; i++ {
+		edit := &metadata.Edit{}
+		kind, picker, op, data := args[0], args[1], args[2], args[3]
+		args = args[4:]
+		switch strings.ToLower(kind) {
+		case "project":
+			edit.Target = &metadata.Edit_Project{
+				Project: &pfs.ProjectPicker{
+					Picker: &pfs.ProjectPicker_Name{
+						Name: picker,
+					},
+				},
+			}
+		default:
+			errors.JoinInto(&errs, errors.Errorf("arg set %d: unknown object type %q", i, kind))
+			continue
+		}
+		switch strings.ToLower(op) {
+		case "add":
+			k, v := parseKV(data)
+			edit.Op = &metadata.Edit_AddKey_{
+				AddKey: &metadata.Edit_AddKey{
+					Key:   k,
+					Value: v,
+				},
+			}
+		case "edit":
+			k, v := parseKV(data)
```

Comment:
Should `pachctl metadata edit $THING edit '{"foo": "bar", "baz": "quux"}'` be allowed, to edit `foo` and `bar` only if they already exist and error otherwise?

`pachctl edit $THING edit …` is a bit infelicitous, but it’s okay.

Reply:
CORE-2231

Reply:
`update` maybe?

Reply:
Maybe we actually want some conditional operations in the pipeline...